sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala

```diff
@@ -1955,20 +1959,29 @@ class Analyzer(override val catalogManager: CatalogManager)
    * @see https://issues.apache.org/jira/browse/SPARK-19737
    */
   object LookupFunctions extends Rule[LogicalPlan] {
+    import CatalogV2Implicits._
     override def apply(plan: LogicalPlan): LogicalPlan = {
       val externalFunctionNameSet = new mutable.HashSet[FunctionIdentifier]()
       plan.resolveExpressions {
-        case f: UnresolvedFunction
-          if externalFunctionNameSet.contains(normalizeFuncName(f.name)) => f
-        case f: UnresolvedFunction if v1SessionCatalog.isRegisteredFunction(f.name) => f
-        case f: UnresolvedFunction if v1SessionCatalog.isPersistentFunction(f.name) =>
-          externalFunctionNameSet.add(normalizeFuncName(f.name))
+        case f @ UnresolvedFunction(NonSessionCatalogAndIdentifier(_, _), _, _, _, _) =>
```

Comment:
This makes me feel that it's clearer to write if else, which also saves a lot of `AsFunctionIdentifier.unapply` calls.
```
case f @ UnresolvedFunction(CatalogAndIdentifier(catalog, v2Ident), _, ...) =>
  if (CatalogV2Util.isSessionCatalog(catalog)) {
    // do nothing for v2 catalog, as we don't have api to check v2 function existence
  } else {
    val ident = v2Ident.asFunctionIdentifier
    if (...) else if ... else ...
  }
```

Reply:
Good point. I've combined these clauses into one. Actually, I think it may be better to add the `v1SessionCatalog.isRegisteredFunction(FunctionIdentifier(name))` into `AsFunctionIdentifier`, so that we can just have this logic:
```
case f @ UnresolvedFunction(AsFunctionIdentifier(ident), ...) =>
  // process V1 session catalog function

case f @ UnresolvedFunction(nameParts, ...) =>
  // process V2 catalog function
```

Let me know what you think.